src/nvim/os/fs.c

```diff
@@ -429,6 +430,64 @@ int os_open(const char *path, int flags, int mode)
   return r;
 }
 
+/// Compatibility wrapper conforming to fopen(3).
+///
+/// Windows: works with UTF-16 filepaths by delegating to libuv (os_open).
+///
+/// Future: remove this, migrate callers to os/fileio.c ?
+///         But file_open_fd does not support O_RDWR yet.
+///
+/// @param flags
+/// @param mode
+/// @return
+FILE *os_fopen(const char *path, const char *flags)
+{
+  assert(flags != NULL && strlen(flags) <= 2);
+  int iflags = 0;
+  // Per table in fopen(3) manpage.
+  if (flags[1] == '\0' || flags[1] == 'b') {
+    switch (flags[0]) {
+      case 'r':
+        iflags = O_RDONLY;
+        break;
+      case 'w':
+        iflags = O_WRONLY | O_CREAT | O_TRUNC;
+        break;
+      case 'a':
+        iflags = O_WRONLY | O_CREAT | O_APPEND;
+        break;
+      default:
+        abort();
+    }
+#ifdef WIN32
+    if (flags[1] == 'b') {
+      iflags |= O_BINARY;
+    }
+#endif
+  } else {
+    // char 0 must be one of ('r','w','a').
+    // char 1 is always '+' ('b' is handled above).
+    assert(flags[1] == '+');
+    switch (flags[0]) {
+      case 'r':
+        iflags = O_RDWR;
+        break;
+      case 'w':
+        iflags = O_RDWR | O_CREAT | O_TRUNC;
+        break;
+      case 'a':
+        iflags = O_RDWR | O_CREAT | O_APPEND;
+        break;
+      default:
+        abort();
+    }
+  }
+  // Per open(2) manpage.
+  assert((iflags|O_RDONLY) || (iflags|O_WRONLY) || (iflags|O_RDWR));
+  int fd = os_open(path, iflags, 0666);
```

Comment:
seems `os_open` will return negative value on error. Perhaps explicitly return NULL then?

Reply:
For reference: returning NULL is enough, because all existing calls to `mch_fopen` don't check `errno`.